agents/supervisor.py

```diff
@@ -0,0 +1,179 @@
+        """Starts the event loop. Blocks until excplicitly stopped."""
+        if not self.coros:
+            maybe_log_message(
+                'Coroutine queue is empty',
+                logger=self.logger,
+                level=logging.WARNING,
+            )
+
+        coro.event_loop(timeout)
+
+    def sleep(self, interval):
+        """
+        Sleep and yield to event loop for a duration of the interval.
+
+        interval (int): Sleep interval duration (in seconds).
+        """
+        if interval < 0:
+            maybe_log_message(
+                'Negative sleep interval not allowed', logger=self.logger
+            )
+            return
+
+        coro.sleep_relative(interval)
+
+    def schedule(
+        self,
+        task,
+        max_retries=3,
+        interval=5,
+        timeout=None,
+        weak=False,
+        *args,
+        **kwargs
+    ):
+        """
+        Schedule a periodic coroutine task.
+
+        Parameters:
+            task (Callable): Function-like to execute periodically.
+            max_retries (int, optional): Maximum number of retries on
+                failure. Default is 3.
+            interval (int, optional): Time (in seconds) between retries.
+                Default is 5.
+            timeout (int, optional): Timeout for task run. Task is
+                considered timed out if it didn't execute in the
+                allocated time. Default is None.
+            weak (bool, optional): 'Weak' rask will not be tracked by the
+                supervisor, i.e., the event loop can be stopped regardless
+                of whether the task has finished. Default is False.
+            *args: Positional arguments passed to task's callable.
+            **kwargs: Keyword arguments passed to task's callable.
+        """
+        idx = self.last_coro + 1
+
+        def run_task():
+            try:
+                for retry in range(1, max_retries + 1):
+                    try:
+                        if timeout:
+                            coro.with_timeout(timeout, task, *args, **kwargs)
+                        else:
+                            task(*args, **kwargs)
+                    except coro.TimeoutError:
+                        maybe_log_message(
+                            'Task %d timed out' % idx, logger=self.logger
+                        )
+                    except Exception as e:
+                        maybe_log_message(
+                            'Task %d failed due to error: %s' % (idx, str(e)),
+                            logger=self.logger,
+                        )
+                    else:
+                        maybe_log_message(
+                            'Task %d finished successfully' % idx,
+                            logger=self.logger,
+                            level=logging.INFO,
+                        )
+
+                        return
+
+                    if retry != max_retries:
+                        self.sleep(interval)
```

Comment:
Optionally support exponential backoff or jitter-based delay (`sleep(interval * retry)` or random delay). This helps avoid thundering herd problems if multiple agents fail simultaneously.

Reply:
Based on [this article](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/), I've ended up using decorrelated jitter backoff (implemented as a generator).

```python
def jitter(base, cap):
    """
    Generator for decorralated jitter backoff.

    Parameters:
        base (int | float): Minimum delay (in seconds).
        cap (int | float): Maximum delay (in seconds).
    """
    interval = base

    while True:
        interval = min(cap, random.uniform(base, interval * 3))
        yield interval
```